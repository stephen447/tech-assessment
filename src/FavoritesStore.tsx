"use client";
import { makeAutoObservable } from "mobx";

interface Token {
  id: string;
  name: string;
  symbol: string;
}

class FavoritesStore {
  favorites: Token[] = [];

  constructor() {
    makeAutoObservable(this);
    if (typeof window !== "undefined") {
      this.favorites = JSON.parse(localStorage.getItem("favorites") || "[]");
    }
  }

  addFavorite(token: { name: string; symbol: string }) {
    if (!this.favorites.find((t) => t.name === token.name)) {
      //  Get highest id and increment by 1
      const id = Math.max(...this.favorites.map((t) => parseInt(t.id))) + 1;
      this.favorites.push({ id: id.toString(), ...token });
    }
    // Update the local storage
    localStorage.setItem("favorites", JSON.stringify(this.favorites));
  }

  removeFavorite(name: string) {
    console.log("Removing favorite with id: ", name);
    this.favorites = this.favorites.filter((token) => token.name !== name);
    // Update the local storage
    localStorage.setItem("favorites", JSON.stringify(this.favorites));
  }

  checkIfFavorite(name: string) {
    return this.favorites.some((token) => token.name === name);
  }
}

export const favoritesStore = new FavoritesStore();
