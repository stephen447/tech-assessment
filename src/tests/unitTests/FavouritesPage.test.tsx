import { render, screen } from "@testing-library/react";
import { favoritesStore } from "../../FavoritesStore";
import FavoritesPage from "../../app/favourites/page"; // Adjust based on your file structure
// import { useRouter } from "next/router";

// Mock the Next.js `useRouter` hook
// jest.mock("next/router", () => ({
//   useRouter: jest.fn(),
// }));

jest.mock("../../components/FavouriteItem", () => {
  return {
    __esModule: true,
    default: ({
      token,
      onRemove,
    }: {
      token: { name: string };
      onRemove: () => void;
    }) => (
      <div>
        {token.name}
        <button onClick={onRemove}>Remove</button>
      </div>
    ),
  };
});

jest.mock("../../FavoritesStore", () => ({
  favoritesStore: {
    favorites: [],
    removeFavorite: jest.fn(),
  },
}));

describe("FavoritesPage", () => {
  //   let mockPush: jest.Mock;

  //   beforeEach(() => {
  //     mockPush = jest.fn();
  //     (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
  //   });

  // Test 1: Render "No favorites added yet" when the favorites list is empty
  it("renders 'No favorites added yet' when the favorites list is empty", () => {
    favoritesStore.favorites = []; // Ensure favorites is empty
    render(<FavoritesPage />);

    const noFavoritesMessage = screen.getByTestId("no-favorites");
    expect(noFavoritesMessage).toBeInTheDocument();
  });

  // Test 2: Render a list of favorite items when favorites are present
  it("renders a list of favorite items", () => {
    favoritesStore.favorites = [
      { id: "1", name: "Favorite 1" },
      { id: "2", name: "Favorite 2" },
    ];

    render(<FavoritesPage />);

    expect(screen.getByText("Favorite 1")).toBeInTheDocument();
    expect(screen.getByText("Favorite 2")).toBeInTheDocument();
  });

  // Test 3: Calls removeFavorite method when a favorite is removed
  //   it("removes a favorite item when the remove button is clicked", () => {
  //     favoritesStore.favorites = [
  //       { id: "1", name: "Favorite 1" },
  //       { id: "2", name: "Favorite 2" },
  //     ];

  //     render(<FavoritesPage />);

  //     // Assuming each favorite item has a remove button, find the first one
  //     const removeButton = screen.getAllByRole("button")[0]; // Adjust if your button has a specific role
  //     fireEvent.click(removeButton);

  //     // Ensure the removeFavorite method was called with the correct argument
  //     expect(favoritesStore.removeFavorite).toHaveBeenCalledWith("1");
  //   });

  // Test 4: Snapshot test to ensure UI consistency
  it("matches the snapshot when favorites are available", () => {
    favoritesStore.favorites = [
      { id: "1", name: "Favorite 1" },
      { id: "2", name: "Favorite 2" },
    ];

    const { asFragment } = render(<FavoritesPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
