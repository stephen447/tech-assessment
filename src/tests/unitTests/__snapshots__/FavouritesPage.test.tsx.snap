// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesPage matches the snapshot when favorites are available 1`] = `
<DocumentFragment>
  <div>
    <header
      class="flex items-center justify-between p-4 bg-background_light dark:bg-background_dark shadow-lg border-b-4 border-border_light dark:border-border_dark h-[10%]"
      role="banner"
    >
      <a
        aria-label="Click to go to homepage"
        class="flex-shrink-0 focus:ring-4"
        href="/"
      >
        <svg
          data-testid="logo"
          height="50"
          viewBox="0 0 100 100"
          width="50"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="50"
            cy="50"
            fill="white"
            r="40"
          />
          <text
            dominant-baseline="middle"
            fill="black"
            font-family="Arial"
            font-size="32"
            font-weight="bold"
            text-anchor="middle"
            x="50%"
            y="55%"
          >
            CS
          </text>
        </svg>
      </a>
      <h1
        class="text-3xl font-bold flex-grow text-center"
      >
        Favourites
      </h1>
      <a
        aria-label="View favorites"
        class="relative px-3"
        data-testid="favorites-link"
        href="/favourites"
        role="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-heart text-3xl text-red-500 hover:text-red-700 transition cursor-pointer"
          data-icon="heart"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
            fill="currentColor"
          />
        </svg>
      </a>
      <button
        class="focus:ring-4 border-2 border-border_light dark:border-border_dark rounded-lg p-2 hover:bg-background_light dark:hover:bg-background_dark transition"
      >
        Light Mode
      </button>
    </header>
    <div>
      <div
        class="flex justify-between items-center m-2"
      >
        <div>
          Favorite 1
          <button>
            Remove
          </button>
        </div>
      </div>
      <div
        class="flex justify-between items-center m-2"
      >
        <div>
          Favorite 2
          <button>
            Remove
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
